### DISCUSSION TITLE
Segmentation fault while running modelfinder in mpi mode

### DISCUSSION THREAD

**Discussion by ycsong** [iqtree/iqtree2#188] [General]
Hello developers,

I am trying to run modelfinder in combination with mpirun, and have been facing some issues. I have a script that I submit to slurm system in our computer here at the lab, and after a couple of minutes, I get the following error:

```
ERROR: PERROR: S--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 15 with PID 106762 on node ta2 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```

The script that I am trying to run is show here:

```
#!/bin/bash
#SBATCH --time=168:00:00
#SBATCH --qos maw
#SBATCH --partition analysis
#SBATCH --nodes=3
#SBATCH --ntasks=36
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=email
#SBATCH --account=account
#SBATCH --job-name model_finder
#SBATCH --error my_job_name-%j.err
#SBATCH --output my_job_name-%j.out

source /etc/profile.d/modules.sh

module purge
module load gcc
module load cmake/3.19.5-intel
module load openmpi


#set +u;source activate metagem;set -u
export PATH="/tahoma/emsls60141/1000soils_data/iqtree2/build:$PATH"

mpirun --bind-to core --map-by core -report-bindings iqtree2-mpi -s ./high_top_gtdbtk_align/align/gtdbtk.bac120.msa.trim.fasta -m TESTONLY -madd LG4M,LG+C10,LG+C20,LG+C30,LG+C40,LG+C50,LG+C60,C10,C20,C30,C40,C50,C60 -safe
```
From what I know, segmentation fault results from not enough memory, but I am not certain where to start the diagnosis process. Your help would be much appreciated.

Thank you.

Young

**Comment by thomaskf** [comment]
Hi Young,

Thanks for reporting this.

In the current releases, the mpi version for ModelFinder is not available
yet.

In fact, we have implemented this, and it is under final testing. It will
be available shortly, say in one or two months. If you really need to use
this feature, you may download the source codes from the "onnxupdate"
branch of our IQTREE2 GitHub page (
https://github.com/iqtree/iqtree2/tree/onnxupdate), compile the source
codes and try it.

Please let me know if you have any questions or if you want to have the
binary version.

Thanks,
Thomas

**Comment by ycsong** [comment]
Hi Thomas,

Many thanks for the update. I think at this point, our colleagues and I would like to give the "onnxupdate" a try and monitor the situation. 

Thanks again,

Young

**Comment by thomaskf** [comment]
Hi Young,

This is just a note that there was an update on the branch "onnxupdate"
today. We fixed an issue that happens when using hybrid mode—openmp and mpi
together. If you are using hybrid mode, please download the updated
version. If you are only using mpi, then it should be fine.

Thanks,
Thomas

**Comment by ycsong** [comment]
Thanks for the update Thomas, it is much appreciated.

Is there an installation guideline somewhere for this? I tried looking at README, but wasn't able to find anything there.

Thank you,

Young

**Comment by thomaskf** [comment]
Hi Young,

You may refer to the section "Compiling MPI version" on our wiki page
regarding the compilation guide:
https://github.com/iqtree/iqtree2/wiki/Compilation-Guide

Let me briefly explain here.

First, you need to have the mpi installed on the machine.
For example, in MacBook:

brew install openmpi


Then, you can get the source code of the branch "onnxupdate" (as you want
to try the MPI version of ModelFinder):

git clone --recursive https://github.com/iqtree/iqtree2
cd iqtree2
git checkout onnxupdate


Then, you can compile the code:

mkdir build
cmake -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx ..
make -j


Sometimes, the openmp include and lib folders cannot be located. Then you
need to specify their locations:

export LDFLAGS=[location of openmp lib folder] (for example:
"-L/opt/homebrew/opt/libomp/lib")
export CPPFLAGS=[location of openmp include folder] (for example:
"-I/opt/homebrew/opt/libomp/include")
cmake -DCMAKE_CXX_FLAGS="$LDFLAGS $CPPFLAGS" -DCMAKE_C_COMPILER=mpicc
-DCMAKE_CXX_COMPILER=mpicxx ..

make -j


If the source codes can be compiled successfully, then you will find an
executable file called "iqtree2-mpi".

You can then run IQTREE2 by the command:

mpirun -np [# of processors] ./iqtree2-mpi ....



I hope the above guidelines will be helpful to you.

Cheers,

Thomas

**Comment by ycsong** [comment]
Hi Thomas,

Thanks again for your help so far. Unfortunately, it appears that the onnxupdate has similar issues that I encountered previously. I recently received a following error message:


```
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 64 with PID 81051 on node ta6 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```

The slurm code is pretty much the same as the one I provided earlier. 

Thank you,

Young

**Comment by thomaskf** [comment]
Hi Young,

Thanks for reporting this. Is it possible to send me the log file, the data file, and the partition file for checking the issue?

Thanks,
Thomas

From: ycsong ***@***.***>
Date: Wednesday, 8 May 2024 at 12:52 AM
To: iqtree/iqtree2 ***@***.***>
Cc: Thomas Wong ***@***.***>, Comment ***@***.***>
Subject: Re: [iqtree/iqtree2] Segmentation fault while running modelfinder in mpi mode (Discussion #188)

Hi Thomas,

Thanks again for your help so far. Unfortunately, it appears that the onnxupdate has similar issues that I encountered previously. I recently received a following error message:

Primary job  terminated normally, but 1 process returned

a non-zero exit code. Per user-direction, the job has been aborted.

--------------------------------------------------------------------------

--------------------------------------------------------------------------

mpirun noticed that process rank 64 with PID 81051 on node ta6 exited on signal 11 (Segmentation fault).

--------------------------------------------------------------------------

The slurm code is pretty much the same as the one I provided earlier.

Thank you,

Young

**Comment by ycsong** [comment]
Hi Thomas,

Apologies for the delayed response. I've attached the log file and the modelwq.gz file that was generated during the execution. I tried to load the fasta file but it's too big for the attachment.

To explain everything from the beginning. At very first, I ran the non-mpi version of ModelFinder. It was working, but was running slowly (i.e., finished testing close to 30 out of 97 total models in a week***, with 120GB RAM provided). Hence I decided to explore a possibility of the MPI version.

*** In our internal server, we can only provide upto a maximum runtime of a week per job, so the non-mpi execution stopped after this period. I am aware that if I were to run this again, IQ-TREE would pick things up from where it previously ended, but thought I look into more efficient option.

Thanks again, and please let me know if you require anything else to make this process easier. I tried looking for a partition file, but it appears that I don't have one with me (or I can't find it).

[gtdbtk.bac120.msa.trim.fasta.modelwq.gz](https://github.com/iqtree/iqtree2/files/15250592/gtdbtk.bac120.msa.trim.fasta.modelwq.gz)
[gtdbtk.bac120.msa.trim.fasta.log](https://github.com/iqtree/iqtree2/files/15250562/gtdbtk.bac120.msa.trim.fasta.log)

**Comment by ycsong** [comment]
Hi Thomas,

I am reattaching the file, gtdbtk.bac120.msa.trim.fasta.model.gz. I was experimenting with the file, and somehow named it to gtdbtk.bac120.msa.trim.fasta.modelwq.gz.  Thank you.

Young
[gtdbtk.bac120.msa.trim.fasta.model.gz](https://github.com/iqtree/iqtree2/files/15286072/gtdbtk.bac120.msa.trim.fasta.model.gz)

**Comment by thomaskf** [comment]
Thanks, @ycsong.
Just a quick note that we have been working on this. I will let you know when there is any progress.
Thanks,
Thomas

**Comment by thomaskf** [comment]
Hi @ycsong ,
An update from us. We have been working, testing and updating the branch "onnxupdate" for the last few weeks. The latest updated version is more stable. In case if you still want to use the mpi version for ModelFinder, you may try the updated version of the branch "onnxupdate". 
Thanks,
Thomas

**Reply by ycsong** [reply]
Hello Thomas, 

Thank you very much for doing this. I recently started another run of the non-mpi job, but I am certain that I will give the new update a try in the near future. I will provide update as it becomes available.  Thanks again.

Young
